interfaces/repo.go

```diff
@@ -0,0 +1,569 @@
+		return fmt.Errorf("cannot add interface: %q, interface name is in use", interfaceName)
+	}
+	r.ifaces[interfaceName] = i
+	return nil
+}
+
+// AllPlugs returns all plugs of the given interface.
+// If interfaceName is the empty string, all plugs are returned.
+func (r *Repository) AllPlugs(interfaceName string) []*Plug {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Plug
+	for _, plugsForSnap := range r.plugs {
+		for _, plug := range plugsForSnap {
+			if interfaceName == "" || plug.Interface == interfaceName {
+				result = append(result, plug)
+			}
+		}
+	}
+	sort.Sort(byPlugSnapAndName(result))
+	return result
+}
+
+// Plugs returns the plugs offered by the named snap.
+func (r *Repository) Plugs(snapName string) []*Plug {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Plug
+	for _, plug := range r.plugs[snapName] {
+		result = append(result, plug)
+	}
+	sort.Sort(byPlugSnapAndName(result))
+	return result
+}
+
+// Plug returns the specified plug from the named snap.
+func (r *Repository) Plug(snapName, plugName string) *Plug {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.plugs[snapName][plugName]
+}
+
+// AddPlug adds a plug to the repository.
+// Plug names must be valid snap names, as defined by ValidateName.
+// Plug name must be unique within a particular snap.
+func (r *Repository) AddPlug(plug *Plug) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject snaps with invalid names
+	if err := snap.ValidateName(plug.Snap); err != nil {
+		return err
+	}
+	// Reject plug with invalid names
+	if err := ValidateName(plug.Name); err != nil {
+		return err
+	}
+	i := r.ifaces[plug.Interface]
+	if i == nil {
+		return fmt.Errorf("cannot add plug, interface %q is not known", plug.Interface)
+	}
+	// Reject plug that don'i pass interface-specific sanitization
+	if err := i.SanitizePlug(plug); err != nil {
+		return fmt.Errorf("cannot add plug: %v", err)
+	}
+	if _, ok := r.plugs[plug.Snap][plug.Name]; ok {
+		return fmt.Errorf("cannot add plug, snap %q already has plug %q", plug.Snap, plug.Name)
+	}
+	if r.plugs[plug.Snap] == nil {
+		r.plugs[plug.Snap] = make(map[string]*Plug)
+	}
+	r.plugs[plug.Snap][plug.Name] = plug
+	return nil
+}
+
+// RemovePlug removes the named plug provided by a given snap.
+// The removed plug must exist and must not be used anywhere.
+func (r *Repository) RemovePlug(snapName, plugName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such plug exists
+	plug := r.plugs[snapName][plugName]
+	if plug == nil {
+		return fmt.Errorf("cannot remove plug %q from snap %q, no such plug", plugName, snapName)
+	}
+	// Ensure that the plug is not used by any slot
+	if len(r.plugSlots[plug]) > 0 {
+		return fmt.Errorf("cannot remove plug %q from snap %q, it is still connected", plugName, snapName)
+	}
+	delete(r.plugs[snapName], plugName)
+	if len(r.plugs[snapName]) == 0 {
+		delete(r.plugs, snapName)
+	}
+	return nil
+}
+
+// AllSlots returns all plug slots of the given interface.
```

Comment:
plug slots?  There are several of those.

Can you please do a pass of self-review to make sure the automated renames you've done ended up well?